src/day_3.rs

```diff
@@ -0,0 +1,88 @@
+use std::fs::File;
+use std::io::{self};
+
+// Number of bits for each line in the problem input
+const N_BITS: usize = 12;
+
+// For a sequence of size N_BITS strings of 1s and 0s
+// Return as a tuple
+// - String with 1s in each position where 1 is most common.
+// - String with 0s in each position where 1 is most common.
+fn most_popular_bit_strings(lines: &Vec<String>) -> (String, String) {
+  // Threshold for determining whether '1s' are 'more common'
+  // This is half the total number of lines.
+  let midpoint = lines.len() / 2;
+  // Count of '1's in each position
+  let ones_counts = lines.iter().fold(vec![0; N_BITS], |mut counts, line| {
+    for (index, bit) in line.char_indices() {
+      if bit == '1' {
+        counts[index] += 1
+      }
+    }
+
+    counts
+  });
+
+  let mut gamma = "".to_owned();
+  let mut epsilon = "".to_owned();
+
+  ones_counts.iter().for_each(|n| {
+    if *n >= midpoint {
+      gamma.push('1');
+      epsilon.push('0');
+    } else {
+      gamma.push('0');
+      epsilon.push('1');
+    }
+  });
+
+  return (gamma, epsilon);
+}
+
+/**
+ * Given a vec of Strings, `strings` and a search String `search`,
+ * Starting with the minimal substring of `search`, find a set of
+ * matching Strings from `strings`. Once a single match exists for
+ * a given substring, return it.
+ */
+fn get_only_substring(strings: &Vec<String>, search: String) -> String {
```

Comment:
I feel like you have miss-understood this one. Frequency of bits needs to be re-calculated by the remaining bytes.

As such you won't even need the search string, but you will need something to differentiate between oxygen and CO2 as they break ties differently

Reply:
https://github.com/jling90/advent-of-code-2021/pull/4/commits/286d47aa2d73b5f242ebe4b088b112d89dbd2bf2 we did it fam